fix(mcdu): Fix input issues on PERF/W&B and fuel prediction pages

felixharnstrom opened the pull request:
<!-- ⚠⚠ Do not delete this pull request template! ⚠⚠ -->
<!-- Pull requests that do not follow this template are likely to be ignored. -->

<!-- Add the issues this PR fixes here. If no issues are related to this PR, then this line can be removed. -->
<!-- Add further issues with a full "Fixes #[issue_no]" line to ensure GitHub closes each one when the PR is merged. -->
Fixes #3525
Fixes #3321
Fixes #3528

## Summary of Changes
<!-- Please provide a summary of changes for this pull request, ensuring all changes are explained. -->
Fixed a number of missing or incorrect unit conversions relating to fuel and payload weight conversions when using imperial units. This fixes a number of issues where entered values were either not stored correctly or not displayed correctly. The following fields are fixed:

- PERF/W&B
   -  Page 1
      - BLOCK FUEL
      - TAXI FUEL
      - TRIP FUEL
   - Page 2
     - PAYLOAD
     - FWD BAG
     - REAR BAG
- Fuel prediction pages (INIT page B and the FUEL PRED page)
   - TAXI (on INIT B)
   - RTE RSV/%
   - ALTN/TIME
   - FINAL/TIME
   - MIN DEST FOB
   - EXTRA/TIME

Max allowed block fuel value is also changed to 19046 kg to match MSFS fuel dialog and actual fuel capacity.
Also fixes an issue where TRIP FUEL input on the PERF/W&B page was incorrectly bound to 2R.
## Screenshots (if necessary)
<!-- If your PR includes visual changes, screenshots from before and after your change should always be included. -->
<!-- Please make best efforts to provide useful before and after screenshots. They should match camera angle, zoom, size, time of day, etc. -->
Before             |  After
:-------------------------:|:-------------------------:
![fuel_conversion_bug](https://user-images.githubusercontent.com/5500848/107995742-17a4d280-6fdf-11eb-9e5b-1e8628651d2d.gif) | ![fuel_conversion_fixed](https://user-images.githubusercontent.com/5500848/107995749-1a9fc300-6fdf-11eb-9aa0-c09e548fa998.gif)
![payload_conversion_bug](https://user-images.githubusercontent.com/5500848/107993418-ae22c500-6fda-11eb-9714-6d0f9b7d0c5a.gif) | ![payload_conversion_fixed](https://user-images.githubusercontent.com/5500848/107993048-53896900-6fda-11eb-99cd-3697341d477d.gif)
![fuel_pred_bug](https://user-images.githubusercontent.com/5500848/107994633-55a0f700-6fdd-11eb-84e3-ce8c6f20ba54.gif) | ![fuel_pred_fixed](https://user-images.githubusercontent.com/5500848/107995008-1d4de880-6fde-11eb-9345-50571e47b762.gif)

## References
<!-- You should be making changes based on some kind reference (manuals, videos, IRL photos). P3D/xplane/fsx references will only be accepted if we believe that one cannot reasonably obtain a better source. Please post screenshots of the references you used. Ask around in the discord for how to find references for what you are working on. Exceptions will probably be made for IRL A320 pilots and engineers. -->

Max fuel capacity from sources in #2286, adjusted to 19046 kg to account for rounding in MSFS

<!-- If you are making a pull request related to the MCDU, please make sure you are ONLY referencing the Honeywell Pegasus Step 1A (Rev 0), 2009 edition manual. -->
<!-- If you do not have this manual, please ask on our discord for assistance -->

## Additional context
<!-- Add any other context about the pull request here. -->

<!-- You may optionally provide your discord username, so that we may contact you directly about the issue. -->
Discord username (if different from GitHub): Jethro#4431

## Testing instructions
<!-- Detail how this PR should be tested by QA. Try to list important items that need checking, either directly changed by this PR or that could be affected by it -->

Test this PR with both imperial and metric units (MCDU MENU -> OPTIONS -> FMGC -> WEIGHT UNIT, make sure to reload the aircraft after changing units): 
1. Make sure that fuel and payload values on PERF/W&B (MCDU MENU -> ATSU -> AOC MENU -> PERF/W&B) are entered correctly and that the correct amount of fuel is loaded when requesting refueling
2. Enter FROM/TO and ALTN on INIT page A 
3. Verify that both fuel prediction pages (INIT page B and FUEL PRED) behave as expected when changing field values

Note that this PR is focused around input validation and unit conversions (with some corresponding improvements to the presentation on the fuel prediction pages). Please bear this in mind when reporting potential issues with calculated values or any underlying algorithms. Also note that I have attempted to make minimal changes to the PERF/W&B payload page since this will be overhauled in an upcoming PR.

<!-- DO NOT DELETE THIS -->
## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created and uploaded.
The build script will have already been run with the latest changes, so no need to rerun it once you download the zip.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the right side, click on the **Artifacts** drop down and click the **A32NX** link

donstim commented:
Thank you for proposing a fix to this issue. There are still issues with units conversion, and it is difficult to tell exactly what MSFS uses in each case. See the following screenshots for example.  First, the MSFS fuel capacities are given in US gallons in the flight_model.cfg file and are reproduced in the MSFS weight and balance window:
![Screenshot (656)](https://user-images.githubusercontent.com/70166617/108006204-b3b6e580-6faf-11eb-9f73-760503133f9c.png)
Showing the same max fuel amount in lbs:
![Screenshot (655)](https://user-images.githubusercontent.com/70166617/108006256-cfba8700-6faf-11eb-8614-7a5aeb48e1c3.png)
MSFS supposedly uses 6.7 lb/gal for fuel density. However, if one does it on a tank by tank basis, the density appears to be 6.69999541 for the center tank, 6.699994493 for the inner tanks, and 6.69995614 for the outer tanks. Now we get to liters:
![Screenshot (657)](https://user-images.githubusercontent.com/70166617/108006688-0349e100-6fb1-11eb-8866-8198877ddf1e.png)
For the center tank, the conversion appears to be 3.784997705 l/gal, right on 3.785 l/gal for the outers, and 3.78495614 l/gal for the inners.  And finally, in kg:
![Screenshot (658)](https://user-images.githubusercontent.com/70166617/108007329-920b2d80-6fb2-11eb-8fc1-3fbc1157b0da.png)
For the center tank, the density is 0.802924761 kg/l, for the inners 0.802923084 kg/l, and for the inners 0.802924783 kg/l.

I'm not sure how to make sense of all this, but the constant for max allowable fuel of 21273 kg is incorrect. As can be seen from the last screenshot, the max allowable fuel in kg should be 19,046 kg. In pounds, it should be 41,989 lb (screenshot 2). Even when I have the units set to lbs in the MCDU options, the max fuel weight value this PR allows me to insert is 21,273.

felixharnstrom commented:
I agree that the 21273 kg fuel limit feels a bit odd. I have actually already opened a separate issue about that here: #3528

Regarding your last point, are you saying that it does not allow you to enter more than 21273 even when set to use LBs? That seems odd to me and not really in line with what I've seen when testing out the PR on my end. For reference, here is what I'm seeing with this PR applied:

KG             |  LB
:-------------------------:|:-------------------------:
![max_fuel_kg_fixed](https://user-images.githubusercontent.com/5500848/108027751-f7026a00-702a-11eb-9b93-5ea2edb4f0eb.gif) | ![max_fuel_lb_fixed](https://user-images.githubusercontent.com/5500848/108027760-fa95f100-702a-11eb-8e57-abffcb675b39.gif)

As you can see, the limit is 21273 when using KGs and 46898 when using LBs.

Did you make sure to reload the aircraft after changing it to use LBs? If not, the system is still in metric until you do. Let me know if you find the time to try it out and see if it's still an issue after reloading the aircraft.

donstim commented:
Ok, I remember that I did not reload the airplane. Thanks for getting back to me.

MisterChocker commented:
Whats the status of this?

felixharnstrom commented:
> Whats the status of this?

I would say that this is ready for QA. I was under the impression that someone else needed to label this as ready for review first though.

MisterChocker commented on A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/ATSU/A320_Neo_CDU_AOC_OFPData.js:
```diff
@@ -137,34 +138,32 @@ class CDUAocOfpData {
-            const maxAllowableFuel = 21273;
-            const enteredFuel = Math.round(+value);
+            const enteredFuel = Math.round(+value) / mcdu._conversionWeight;
             if (enteredFuel >= 0 && enteredFuel <= maxAllowableFuel) {
```
isn't `maxAllowableFuel` undefined at this stage?

felixharnstrom replied:
My understanding was that arrow functions 'inherit' their parent's scope (see line 178 where we access `currentBlockFuel` for instance): https://medium.com/@iampika/es6-arrow-functions-syntax-and-lexical-scoping-d061732071e7

However, I'm not a Javascript expert so I might be misunderstanding something. If we're not sure if it's safe I can certainly revert this.

MisterChocker replied:
I see, I am not an JS expert either, so I'll wait for @beheh response.

beheh replied:
Yeah, it's inherited downwards, so this code is fine.

donstim commented:
> I agree that the 21273 kg fuel limit feels a bit odd. I have actually already opened a separate issue about that here: #3528
> 
So there will be a different PR to fix #3528 ?

felixharnstrom commented:
> > I agree that the 21273 kg fuel limit feels a bit odd. I have actually already opened a separate issue about that here: #3528
> 
> So there will be a different PR to fix #3528 ?

Yes. I don't think it should be done in this PR at least - I would prefer keeping this PR focused on just the unit conversions.
That issue is also somewhat blocked until someone with a better idea of what the max value should actually be is able to weigh in.

donstim commented:
> > So there will be a different PR to fix #3528 ?
> 
> Yes. I don't think it should be done in this PR at least - I would prefer keeping this PR focused on just the unit conversions.
> That issue is also somewhat blocked until someone with a better idea of what the max value should actually be is able to weigh in.

Okay, but the max values are clear in the screenshots I provided above.  Those screenshots show what the max values should be.  They are correct (other than the apparent Asobo bug regarding unusable fuel).

felixharnstrom commented:
> > > So there will be a different PR to fix #3528 ?
> > 
> > 
> > Yes. I don't think it should be done in this PR at least - I would prefer keeping this PR focused on just the unit conversions.
> > That issue is also somewhat blocked until someone with a better idea of what the max value should actually be is able to weigh in.
> 
> Okay, but the max values are clear in the screenshots I provided above. Those screenshots show what the max values should be. They are correct (other than the apparent Asobo bug regarding unusable fuel).

Ah, gotcha. I was just a bit hesitant since I wasn't sure if 19049 (or 19051 according to your PR #2286, this might be due to rounding differences?) was what the value _should_ be. Then, unless you feel that it's better to open a separate PR, I wouldn't mind fixing that while I'm already poking around at this.

donstim commented:
> > > > So there will be a different PR to fix #3528 ?
> > > 
> > > 
> > > Yes. I don't think it should be done in this PR at least - I would prefer keeping this PR focused on just the unit conversions.
> > > That issue is also somewhat blocked until someone with a better idea of what the max value should actually be is able to weigh in.
> > 
> > 
> > Okay, but the max values are clear in the screenshots I provided above. Those screenshots show what the max values should be. They are correct (other than the apparent Asobo bug regarding unusable fuel).
> 
> Ah, gotcha. I was just a bit hesitant since I wasn't sure if 19049 (or 19051 according to your PR #2286, this might be due to rounding differences?) was what the value _should_ be. Then, unless you feel that it's better to open a separate PR, I wouldn't mind fixing that while I'm already poking around at this.

Yeah, it's MSFS rounding differences. Gotta go with MSFS value for it since the units conversion is in the code. The base value in the cfg file that we have control over is in gallons. That is then converted by MSFS for the other units -- lbs, liters, and kgs.

felixharnstrom commented:
> Yeah, it's MSFS rounding differences. Gotta go with MSFS value for it since the units conversion is in the code. The base value in the cfg file that we have control over is in gallons. That is then converted by MSFS for the other units -- lbs, liters, and kgs.

Alright, thanks for the info :) I updated the max allowed fuel to match MSFS.

wpine215 commented:
Is this ready for QA testing again?

felixharnstrom commented:
> Is this ready for QA testing again?

Personally I think so - there is nothing else I would like to add to the PR at least :)
But since this is my first contribution to this project I'm not sure who has the final word on that. I'm guessing only members of the flybywiresim organization?

Benjozork reviewed (approved):
LGTM

marcman86 commented:
Category: QA Tester
Name: marcman86#4907
Date of testing: 11.3.2021
Version of the sim: 1.14.5.0
PR Tested: #3545
Tier of testing: 1
Changes to observe: Fuel and payload behaviour
Testing technique: Tryout

Steps made to test the Pull Request:

1.) Spawned C&D at EDDP
2.) Turned on bat. and ext. power, set unit to kg
3.) Played with PERF/W+B
4.) Played with Init B

Media:

![130](https://user-images.githubusercontent.com/78707844/110858662-a6450080-82ba-11eb-93d5-8c4ef478d65d.png)
Image1

Issues: 
- On small AOC Payload FWD BAG can get negative (Image1)
- FS Fuel page baggage gets not updated
- Fuel gets burned on only bat. and ext. power, about 40kg a minute


Overall Rating: Needs to be improved

Conclusions of the testing: Lots of stuff is working. Small issues found.

Hope this helps :-)

DavidEGTE commented:
Category: QA Tester
Name: David#2177
Date of testing: 11/03/2021
Version of the sim: 1.14.5.0
PR Tested: #3545 
Tier of testing: 1

Changes to observe:
Fuel and payload behaviour

Testing technique: Try-out

Steps you've made to test the Pull Request:
1. Spawned C&D @ LIRN
2. Turned on batt + lights
3. Tested Fuel Pred 
4. Tested Perf/W+B
5.

Flight notes (When done a full flight):
N/A
Media: Insert screenshots, recordings, or anything relevant to the review here.
N/A

Issues (If Any): Write any issues
same as @marcman86 for point 1&2, point 3 I don't see fuel consumption, tried with fuel pump on and off
(tried for 45min spawn since I was watching a series in the meanwhile)
Steps to achieve the issue:
N/A

Overall Rating: To be improved
Conclusions of the testing: Most of the stuff mentioned are working, same needs to be recheck

marcman86 commented:
To get the "fuel leak":

1.) Spawned C&D at EDDP
2.) Turned on bat. and ext. power.
3.) Turned on fuel pumps

4.) Set AOC block fuel to 100, leave taxi and trip fuel open
Here you need watch ECAM fuel page
5.) Clicked on Refuel load
ECAM shows outer tanks 40 and FOB 20kg. Just wait a little and is goes all down to 0

6.) Set AOC block fuel to 1000, leave taxi and trip fuel open
7.) Clicked on Refuel load
ECAM shows outer tanks 480 and FOB 920kg. Just wait a little and is goes down

8.) Set AOC block fuel to 4000, leave taxi and trip fuel open
9.) Clicked on Refuel load
ECAM shows outer tanks 680, inner tanks 1280 and FOB 3920kg. Just wait a little and is goes down

10.) Set AOC block fuel to 5000, leave taxi and trip fuel open
11.) Clicked on Refuel load
ECAM shows outer tanks 680, inner tanks 1780 and FOB 4920kg. Here it gets stable!

felixharnstrom commented:
@marcman86  Thanks for investigating this issue :) I just verified that the same behavior exists when setting the fuel quantity via the MSFS fuel dialog as well, so it's not an input issue with the MCDU. I'm guessing it might be modelling fuel flowing from the tanks to the engines, but I might be wrong about that. Either way I think changing this is outside the scope of this PR since it relates to the underlying fuel consumption rather than being an input issue. Also, I'm guessing this might change with the upcoming changes to the engine model.

felixharnstrom commented:
@marcman86 @DavidEGTE Thank you both for reviewing this. Regarding your other comments:

- The negative payload weights seem to be an existing issue: #3776. So since this is not directly caused by any of the changes made in this PR I would prefer not fixing this here since the balancing algorithm will change fairly drastically in #3474 and I would prefer not doing any more substantial changes that will just be obsolete soon (the input handling is one of the few things that might get reused) :)
- The payload not being updated in the MSFS payload menu seems to be an Asobo bug: https://github.com/flybywiresim/a32nx/pull/1865#issuecomment-745271247. You can verify that the GW value on the ECAM changes and that the suspension gets compressed after requesting the payload.

donstim commented:
> To get the "fuel leak":
> 
This is a known bug in the MSFS outer to inner fuel tank transfer. If the inner tank is below the trigger value, then transfer from the outer tanks will start. However, if the engines are not on, the fuel doesn't actually make it way into the inner tanks. It just disappears.

Once the transfer starts it continues until the airplane is refueled. This is indicated in our airplane by adding enough fuel to an inner tank to give it a quantity of 474 gallons.

St54Kevin commented:
Category: QA Tester
Name: St54Kevin#3583
Date of testing: 19/03/2021
Version of the sim: 1.14.5.0
PR Tested: #3545 
Tier of testing: 1
Changes to observe: MCDU AOC PERF/W&B page

**Steps you've made to test the Pull Request:**
1. Loaded up C&D
2. Tested different kg inputs on AOC and INIT B page
3. changed to lbs, reloaded plane
4. Tested again inputs with lbs

**Media:**
![image](https://user-images.githubusercontent.com/448627/111761856-a63b9680-88a0-11eb-95a1-78c3fd3ddc55.png)

**Issues (If Any):** 
- I can confirm the disappearing fuel with smaller load as mentioned above

**Conclusions of the testing:** LGTM. Overall switching units doesnt seem to cause any issues anymore. Nice fix

FlightSimmer7700 commented:
Category: QA Tester
Name: FlightSimmer7700#3468
Date of testing: 19/3/21
Version of the sim: 1.14.5.0
PR Tested: #3545 
Tier of testing: 1
Changes to observe:
PERF/W&B and fuel prediction

Testing technique: Loaded in and tried
Steps you've made to test the Pull Request:
1. Loaded in
2. filled the PERF/W&B page
3. filled init a and flight plan page
4. filed init b 
5. looked at fuel pred page

Media: Insert screenshots, recordings, or anything relevant to the review here.
![image](https://user-images.githubusercontent.com/70103461/111840766-d90f7a00-88f4-11eb-86ae-ac50217d34e5.png)
![image](https://user-images.githubusercontent.com/70103461/111840793-e167b500-88f4-11eb-9ad8-54aae1114943.png)
![image](https://user-images.githubusercontent.com/70103461/111842664-b6329500-88f7-11eb-9604-a0a7597d4964.png)
https://user-images.githubusercontent.com/70103461/111842857-090c4c80-88f8-11eb-9319-82884e863b0d.mp4
https://user-images.githubusercontent.com/70103461/111840843-f0e6fe00-88f4-11eb-9bc4-3306600a1130.mp4
https://user-images.githubusercontent.com/70103461/111840881-fd6b5680-88f4-11eb-92f3-43d6743a0184.mp4
https://user-images.githubusercontent.com/70103461/111840911-09571880-88f5-11eb-96df-9fb705687dbd.mp4
https://user-images.githubusercontent.com/70103461/111843103-7ddf8680-88f8-11eb-9cea-4f0acc6c1b93.mp4

Issues (If Any): 
- more of a suggestion but maybe make the min you are able to put into the block fuel area the 1% in the msfs menu (currently lets you enter even 1 as shown in the screenshot
- FWD baggage goes into minus when the payload is from 1 to 701
- negative values can be entered in both FBW and rear baggage
- PAX number cannot be entered
- When entering RTV RSV it doesn't enter and shows no message but % is allowed
- When entering a negative into MIN DEST it says FORMAT ERROR but when entered into FINAL or ALTN it says ENTRY OUT OF RANGE. Then when entering a negative into RTE RSV/% it doesn't say anything or enter

Steps to achieve the issue
do as shown in the Media section above

Overall Rating: To be improved
Conclusions of the testing: Quite a few things I have found which may be realistic or how they are supposed to be but just pointing everything that seems wrong to me.

felixharnstrom commented:
@FlightSimmer7700 Thank you for the in-depth review.

- I would prefer not adding any logic to the minimum fuel amount based on the MSFS fuel dialog. I simply don't see any reason why we should limit ourselves to such arbitrary restrictions in the MSFS UI unless we really need to.
- I added a check to make sure that the entered payload values are positive. Other than that I would prefer not to make any more changes to the payload page since it will change fairly drastically in an upcoming PR as I described in a previous comment. That includes the inoperable PAX field and the negative calculated payload values.
- I fixed the error message when entering a negative value in MIN DEST. It should now say ENTRY OUT OF RANGE like the other fields.

I haven't done any changes to the RTE RSV field yet. I'll need to dig around a bit and try to determine how it should behave before I make any changes to the input logic.

FlightSimmer7700 commented:
> @FlightSimmer7700 Thank you for the in-depth review.
> 
> * I would prefer not adding any logic to the minimum fuel amount based on the MSFS fuel dialog. I simply don't see any reason why we should limit ourselves to such arbitrary restrictions in the MSFS UI unless we really need to.
> * I added a check to make sure that the entered payload values are positive. Other than that I would prefer not to make any more changes to the payload page since it will change fairly drastically in an upcoming PR as I described in a previous comment. That includes the inoperable PAX field and the negative calculated payload values.
> * I fixed the error message when entering a negative value in MIN DEST. It should now say ENTRY OUT OF RANGE like the other fields.
> 
> I haven't done any changes to the RTE RSV field yet. I'll need to dig around a bit and try to determine how it should behave before I make any changes to the input logic.

Sounds good, I'll take a look at the pr later today and check that everything is working. Apart from the RTE RSV

FlightSimmer7700 commented:
Ok, I tried it out again. Cannot input negative values but they still come up when putting low payload(I think that will be fixed in the future as you said). I found the tiniest issue which I'm not sure is even worth commenting and then working on but when entering a negative into min dest you get the ENTRY OUT OF RANGE message but then when you clear that the number is still there while when doing it on the others above the ENTRY OUT OF RANGE comes up and then when cleared the number does not show again

https://user-images.githubusercontent.com/70103461/111888320-514c6d00-89d3-11eb-8acb-a735b134679f.mp4

felixharnstrom commented:
@FlightSimmer7700 Good catch! I pushed a fix for that issue - now the value is cleared like for the other fields.

I've also made a number of changes to the RTE RSV/% and FINAL/TIME fields. Both of these fields now:
- Disallow entry in the `XX/YY` format. The Pilot's guide how only one or the input types shuld be allowed at a time, not both at the same time. So now `XX/` (the trailing slash is optional) is treated as weight input, and `/YY` is interpreted as percentage for the route resereve and time for the final fuel.
- Made the input parser a lot stricter overall. For instance, the previous implementation allowed input like `/0.2/ABC`. Now the only allowed input formats are `XX/`, `XX` and `/XX`.
- The manually entered value is more consistently displayed in large font and the calculated value in small font.

I did not change how the RTE RSV/% field is still replaced with all slashes when an out of range value is entered.

Lucky38i reviewed (approved):
LGTM, just needs QA to test all possible entries and void values

felixharnstrom commented:
@MisterChocker This PR will actually not fix the issue that you linked (#3776) 🙂  That is an issue with the underlying payload distribution algorithm: If you enter a small enough payload value, the calculated rear and forward baggage values can be negative. I have not fixed this in this PR since this is not really an issue with the input handling per se and that page will change fairly drastically in #3474 anyway. All I've done in this PR is make sure that the user cannot enter any negative values directly.

TheSquirtingPantin commented:
Hello All, 

Category: QA Trial
Name: Discord User Rémi#8871
Date of testing: Current date 23/03/2021
Version of the sim: V 1.14.5.0
PR Tested: #3545
Tier of testing: 1
Changes to observe:
Logic of PERF/WB and FUEL PRED pages

Testing technique: Description of how you tested the Pull Request
Steps you've made to test the Pull Request:
1. C/D at KDEN
2. Check for void and logic in PERF/WB Page
3. 
4.
5.

Flight notes (When done a full flight):N/A

Media: N/A

Issues (If Any):
	- Entered random value for "Payload" in PERF/WB 2/2 page (12000kg). Then selected OFP SEND REQUEST (from SIMBRIEF with 160 pax). Clicking on 1L is calculating new Payload and calculation for payload done with negative rear bags
	- Separation bags and passengers confusing (why not renaming "passenger load", "bag fwd" and "rear bag". Payload being the sum of the 3? Is it a limitation from MSFS?) 
	- When importing the data from OFP, number of passenger is green (imported data). If entering a random number in 3R, then clicking on the 1L "Payload", the data previously registered in 3R is shown (despite 3R being still locked green).

Overall Rating: To be improved
Conclusions of the testing: NA

felixharnstrom commented:
@TheSquirtingPantin Thank you for the review. As I mentioned in a previous comment, the PERF W/B payload page will be overhauled in an upcoming PR. As such I would prefer not making any more changes to this page than the basic input validation and unit conversions already included in this PR since any such changes will soon be obsolete.

BritishAvgeek commented:
Category: QA Trial
Name: BritishAvgeek#1436
Date of testing: 24/03/2021
Version of the sim: 1.14.5.0
PR Tested: #3545
Tier of testing: 1
Changes to observe: Input changes to the Perf/W&B and Fuel Prediction pages 

Testing technique: 
Spawned C&D and input all data for a routine flight and edited the values multiple times.
Reloaded to lbs to check same inputs as a comparison to see if any issues arise 

Steps you've made to test the Pull Request:
1. Noted all comments above in relation to focus on basic input and validation so some issues as mentioned above were out of the scope of this PR
2. Spawned C&D
3. Imported flight plan from EDDB to EFHK using Simbrief 
4. Configured MCDU appropriately as "routine" to check for any problems 
5. Amended the figures input 
6. Tried to input various values above the max and the PR worked and didn't allow it in
7. Repeated the steps having reloaded with the change to lbs and followed all the above in repeat
8. Changed back to KG and reloaded, worked through the flow again to ensure no issues in the reverse of the data change back to kg


Flight notes (When done a full flight):N/A
Media: N/A

Issues (If Any):
None

Overall Rating: Good
Conclusions of the testing: The PR addresses everything it says it does and I found no issues during my tests noting everything commented on and discussed further up. Thanks!

marcman86 commented:
Category: QA Tester
Name: marcman86#4907
Date of testing: 28.3.2021
Version of the sim: 1.14.6.0
PR Tested: #2545
Tier of testing: 1
Changes to observe: PERF/W&B and fuel prediction page inputs
Testing technique: Tryout

Steps made to test the Pull Request:

1.) Spawned C&D at EDDP
2.) Turned on bat. and ext. power
3.) Tested inputs on PERF/W&B page 1&2
4.) Tested inputs on INIT B page

Media:

![150](https://user-images.githubusercontent.com/78707844/112748129-27d6a700-8fba-11eb-983f-86e7a9a7eb51.png)
Image1

Issues: 

- FWD Bag still gets negative when PAYLOAD < 702
- RTE RSV/% only accepts /YY and /YY.Y
- FINAL /TIME is limited to 0130, but XX/ entry can be higher than 1.9 (Image1)

Overall Rating: Needs to be improved

Conclusions of the testing: Small bugs found

Hope this helps :-)

felixharnstrom commented:
@marcman86 Thank you for the review :)

- The payload page will be overhauled significantly in an upcoming PR and since this is an issue with the underlying algorithm rather than a straight-up input issue, I would prefer not adressing this in this PR.
- I have pushed a fix for the route reserved fuel issue, thanks for spotting this!
- The final fuel weight limitation is intended. It should be between 0-100 even though this doesn't exactly correspond to the 1H30M max time:

![image](https://user-images.githubusercontent.com/5500848/112750152-0a5c0a00-8fc7-11eb-9da1-35e4f22a30d7.png)

marcman86 commented:
Retested after fix:

RTE RSV/% still not working.

It just goes ---.-/--.- 
No Entry out of range or format error message is shown

felixharnstrom commented:
@marcman86 Thanks you for taking another look at it. However, I am unable to reproduce this issue. Could you maybe record a short video when you have time? Also, please make sure that you are trying out the artifact built from the latest commit.

For reference, I downloaded the latest package built from this PR from Github and tried it out. As far as I can tell, the route reserved field behaves as expected:

![rte_rsv_fixed2](https://user-images.githubusercontent.com/5500848/112754367-0d152a00-8fdc-11eb-8306-b7466afe28bc.gif)

Contents of `build_info.json`:
```
{
  "built": "2021-03-28T11:21:06+00:00",
  "ref": "merge",
  "sha": "925d3ab4c603c535a94d74a087fade4baae6fe5b",
  "actor": "felixharnstrom",
  "event_name": "manual"
}
```

marcman86 commented:
I waited till the build was finished. But will try again.

MisterChocker commented:
The build is finished

felixharnstrom commented:
@marcman86 Thank you for taking another look and for being patient with me. Please let me know the steps to reproduce if you find that it's still an issue.

marcman86 commented:
Still not working. Same as before.

1.) Spawned C&D at EDDP
2.) Turned on bat. and ext. power
3.) Init A EDDP/EDDB, CI10, FL160
4.) Init B Autofill ZFW/ZFWCG and did (auto-) Fuel Planing
5.) Enter "2" end Press 3L (RTE RSV/%)

shows ---.-/--.-

felixharnstrom commented:
@marcman86 Thanks for taking another look, now I know what you're referring to :) If 2 tonnes (in your example) correspond to a percentage greater than 15%, replacing the field with dashes is the intended (and pre-existing) behavior.

Does the issue with the route reserved weight being 0.0 all the time seem fixed to you?

marcman86 commented:
Oh, ok. but with no "Out of range" message or something? Will check the other thing.

felixharnstrom commented:
@marcman86 The reference is not explicit on this, but the way I read it is that the field gets dashed and no message is shown in the scratchpad. Maybe @Lucky38i can weigh in on this?

MisterChocker commented:
Alright waiting for luckys response before merging.

felixharnstrom commented:
Actually, I just now saw that IRL A320 pilot "polonski" have posted a super helpful reference video on Discord. It seems that percentage entry outside 0-15 *should* be disallowed and give an out of range message, and if the corresponding percentage when entering the fuel weight manually is out of range it should be allowed and just result in the percentage part being dashed, not the whole field. I'll push a new commit tomorrow that fixes this.

felixharnstrom commented:
I decided that sleep is for the weak and did the changes immediately :)

My latest commit fixes a number of input and display issues on the Init B and the fuel prediction pages:
- Only the route reserved percentage field is now white and dashed when the weight is out of range instead of the whole thing.
- It is now explicitly disallowed to enter route reserved percentages outside the 0-15% range.
- The alternate fuel time is now dashed and white if no alternate destination has been set.
- The alternate fuel weight is now green and pilot entry is disallowed if no alternate destination has been set.
- Extra time is now white and dashed if the extra fuel weight is negative.

References:

*Screenshot showing the INIT B page from this Youtube video:* https://www.youtube.com/watch?v=DP8soiH_4jE&t=122s
![cockpit_view](https://user-images.githubusercontent.com/5500848/112165944-7caba380-8bef-11eb-9bf2-428c132612ea.PNG)

Unfortunately, the reference video recorded by Discord user and IRL A320 pilot *polonski#4524* is too large to attach to this comment. Here is the link to the post in Discord: https://discord.com/channels/738864299392630914/739223532269076561/825622509197852672
Here is a screenshot showing the most important part:
![image](https://user-images.githubusercontent.com/5500848/112769556-0b6e5500-9022-11eb-84a0-5d64af2501d0.png)

*Pegasus Step 1A (Rev 0), 2009 edition manual*
![pilots_guide](https://user-images.githubusercontent.com/5500848/112166277-c6948980-8bef-11eb-87e1-bb2348d2cb93.PNG)
![image](https://user-images.githubusercontent.com/5500848/112769590-30fb5e80-9022-11eb-9246-3ae1b9d770bd.png)
![image](https://user-images.githubusercontent.com/5500848/112769759-f9d97d00-9022-11eb-91b3-a04571533cfb.png)
![image](https://user-images.githubusercontent.com/5500848/112769752-ee865180-9022-11eb-9fcd-3584562594de.png)
![image](https://user-images.githubusercontent.com/5500848/112769733-d7476400-9022-11eb-9b2a-75d7acde9838.png)

marcman86 commented:
![153](https://user-images.githubusercontent.com/78707844/112862641-428e4600-90b6-11eb-8050-35ff78bb76df.png)

marcman86 commented:
Category: QA Tester
Name: marcman86#4907
Date of testing: 29.3.2021
Version of the sim: 1.14.6.0
PR Tested: #3545
Tier of testing: 1
Changes to observe: Init B fixes
Testing technique: Tryout

Steps made to test the Pull Request:

1.) Spawned C&D at EDDP
2.) Start up and Init A EDDP/EDDM
3.) Test Init B

Issues: -

Overall Rating: Good

Conclusions of the testing: Works. Nice work!

Hope this helps :-)

DavidEGTE commented:
Category: QA Tester
Name: David#2177
Date of testing: 01/04/2021
Version of the sim: 1.14.6.0
PR Tested: #3545 
Tier of testing: 1

Changes to observe:
Fix input issues on PERF/W&B and fuel prediction pages

Testing technique: Try-out

Steps you've made to test the Pull Request:
1. Spawned C&D
2. Played with PERF/W&B random + max capacity fuel
3. Tested Init Page with some values 

Flight notes (When done a full flight):
N/A
Media: Insert screenshots, recordings, or anything relevant to the review here.
N/A
Issues (If Any): Write any issues
N/A
Steps to achieve the issue:
N/A

Overall Rating: Good
Conclusions of the testing: Works like a charm! GTG

Benjozork commented:
@felixharnstrom this needs to be rebase for hard-fork compatibility.

felixharnstrom commented:
@Benjozork Done!

felixharnstrom commented:
Added a new section in the changelog for the 0.7.0 milestone in my latest commit. Once all the commits for 0.6.0 are in, this should be ready to go :)
